packages/graphiql/src/state/GraphiQLSessionProvider.tsx

```diff
@@ -90,120 +98,117 @@ export const SessionContext = React.createContext<
+        // ...getQueryFacts(schema, value),
+      };
+    }
+    case actionTypes.variables_changed: {
+      return {
+        ...state,
+        variables: {
+          ...state.variables,
+          text: action.payload,
+        },
+      };
+    }
+    case actionTypes.operation_succeeded: {
+      return {
+        ...state,
+        operationErrors: null,
+        results: {
+          ...state.results,
+          text: action.payload.result,
+        },
+      };
+    }
+    case actionTypes.operation_errored: {
+      return {
+        ...state,
+        operationErrors: [String(action.payload)],
+      };
+    }
+    default: {
+      return state;
+    }
+  }
 }
 
 export type SessionProviderProps = {
   sessionId: number;
   fetcher?: Fetcher;
   session?: SessionState;
-  children: any;
+  children: React.ReactNode;
 };
 
 export function SessionProvider({
   sessionId,
   fetcher = defaultFetcher,
   session,
-  ...props
+  children,
 }: SessionProviderProps) {
   const schemaState = React.useContext(SchemaContext);
 
-  const [state, dispatch] = useReducers<SessionState, AT, ActionState>({
-    reducers: [getSessionReducer(schemaState.schema as GraphQLSchema)],
-    init: () => initialState,
-  });
-
-  const operationError = (error: Error) =>
-    dispatch({
-      type: actionTypes.operation_errored,
-      payload: { error, sessionId },
-    });
-  const editorLoaded = (context: EditorContexts, editor: CodeMirror.Editor) => {
-    dispatch({
-      type: actionTypes.operation_errored,
-      payload: { editor, context },
-    });
-  };
-  const changeOperation = (operationText: string) => {
-    dispatch({
-      type: actionTypes.operation_changed,
-      payload: { value: operationText, sessionId },
-    });
-  };
-
-  const changeVariables = (variablesText: string) =>
-    dispatch({
-      type: actionTypes.variables_changed,
-      payload: { value: variablesText, sessionId },
-    });
-  const executeOperation = async (
-    sessionState: SessionState,
-    operationName?: string,
-  ) => {
-    try {
-      dispatch({
-        type: actionTypes.operation_requested,
-      });
-
-      const fetchValues: GraphQLParams = {
-        query: sessionState?.operation?.text ?? '',
-      };
-      if (sessionState.variables?.text) {
-        fetchValues.variables = sessionState.variables.text as string;
-      }
-      if (operationName) {
-        fetchValues.operationName = operationName as string;
+  const [state, dispatch] = React.useReducer(reducer, sessionId, () => ({
+    ...initialState,
+    sessionId,
+  }));
+
+  const operationError = useDispatchedCallback(dispatch, operationErrorAction);
+  const editorLoaded = useDispatchedCallback(dispatch, editorLoadedAction);
+  const changeOperation = useDispatchedCallback(
+    dispatch,
+    changeOperationAction,
+  );
+  const changeVariables = useDispatchedCallback(
+    dispatch,
+    changeVariablesAction,
+  );
+
+  const executeOperation = React.useCallback(
```

Comment:
i'm not as strong with the `useCallback` hooks but would it be possible to instead have the `sessionState` argument being passed in here, we could just have the useCallback watch the props it cares from from the state in this scope? see my comment here on the base branch for more context on what i'm trying to get at:

https://github.com/graphql/graphiql/pull/1359/files/97f347f5946126e75961568b663f918ddb6530ef#r403618312

Reply:
like could we do something like

``` TypeScript
const executeOperation = React.useCallback(
   async (operationName?: string) => {
    // ...
}, [state.operation, state.variables])
```

Reply:
totally agree! I hadn't looked at the signature of this function yet, good call